packages/agent-utils/src/llm/ChatLogs.ts

```diff
@@ -0,0 +1,76 @@
+import { ChatCompletionRequestMessage as ChatMessage } from "openai";
+
+export { ChatMessage };
+
+export type ChatLogType =
+  | "persistent"
+  | "temporary";
+
+export interface ChatLog {
+  tokens: number;
+  msgs: ChatMessage[];
+}
+
+export class ChatLogs {
+  private _logs: Record<ChatLogType, ChatLog> = {
+    "persistent": {
+      tokens: 0,
+      msgs: [],
+    },
+    "temporary": {
+      tokens: 0,
+      msgs: [],
+    },
+  };
+
+  constructor(logs?: Record<ChatLogType, ChatLog>) {
+    if (logs) {
+      this._logs = logs;
+    }
+  }
+
+  get tokens(): number {
+    return this._logs["persistent"].tokens +
+      this._logs["temporary"].tokens;
+  }
+
+  get messages(): ChatMessage[] {
+    return [
+      ...this._logs["persistent"].msgs,
+      ...this._logs["temporary"].msgs
+    ];
+  }
+
+  static from(chatLog: ChatLog): ChatLogs {
+    return new ChatLogs({
+      "persistent": chatLog,
+      "temporary": {
+        tokens: 0,
+        msgs: []
+      }
+    });
+  }
+
+  public get(type: ChatLogType): ChatLog {
+    return this._logs[type];
+  }
+
+  public add(type: ChatLogType, log: ChatLog) {
+    this._logs[type].tokens += log.tokens;
+    this._logs[type].msgs.push(...log.msgs);
+  }
+
+  public clone(): ChatLogs {
+    return new ChatLogs(
+      JSON.parse(JSON.stringify(this._logs))
+    );
+  }
+
+  public toString() {
+    return JSON.stringify(this, null, 2);
+  }
+
+  public toJSON() {
+    return this._logs;
```

Comment:
Shouldn't this be cloned?
Also, toJSON kinda implies it's returning a string IMO, could be wrong
Also, can you add return types here and to `toString()`?

Reply:
No need to clone, that will happen naturally when `JSON` stringifies the object. Also you do not return a string from `toJSON`, it's a custom handler that allows you to return the object you want to be stringified. You can learn more about it here: https://www.w3docs.com/learn-javascript/working-with-json.html

Reply:
Just added return types.

Reply:
Good to know, thanks!